src/Analyzers/CSharp/Tests/ForEachCast/ForEachCastTests.cs

```diff
@@ -1178,5 +1178,80 @@ public void Reset()
             await TestAlwaysAsync(test, test);
             await TestWhenStronglyTypedAsync(test, test);
         }
+
+        [Fact, WorkItem(63470, "https://github.com/dotnet/roslyn/issues/63470")]
+        public async Task WeaklyTypedGetEnumeratorWithIEnumerableOfT_DifferentTypeUsedInForEach()
+        {
+            var code = @"
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+public class C : IEnumerable<string>
+{
+    public IEnumerator GetEnumerator() => new Enumerator(); // compiler picks this for the foreach loop.
+
+    IEnumerator<string> IEnumerable<string>.GetEnumerator() => null; // compiler doesn't use this.
+
+    public static void M(C c)
+    {
+        [|foreach|] (C x in c)
+        {
+        }
+    }
+
+    private class Enumerator : IEnumerator
+    {
+        public object Current => ""String"";
+
+        public bool MoveNext()
+        {
+            return true;
+        }
+
+        public void Reset()
+        {
+        }
+    }
+}
+";
+
+            var fixedCode = @"
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+public class C : IEnumerable<string>
+{
+    public IEnumerator GetEnumerator() => new Enumerator(); // compiler picks this for the foreach loop.
+
+    IEnumerator<string> IEnumerable<string>.GetEnumerator() => null; // compiler doesn't use this.
+
+    public static void M(C c)
+    {
+        foreach (C x in c.Cast<C>())
```

Comment:
Sorry, I don't have this all paged in, but why is it an improvement to call `.Cast()` here?

Reply:
I guess the point is "this is how you suppress teh warning"? Since the IEnumerable<T> impl is not IEnumerable<C>, it doesn't meet what we want/expect to use with `foreach (C x in c...)`